File: 2-8/2-8-ruby/bike.rb
# frozen_string_literal: true

# Class representing a bike, extending the Vehicle class.
# @version 1.0
# @since   2024-11-05

# The Bike class represents a bike, which is a specific type of vehicle
class Bike < Vehicle
  # The cadence of the bike.
  # @type [Integer]
  attr_accessor :cadence

  # The gear of the bike.
  # @type [Integer]
  attr_accessor :gear

  # Creates a new Bike instance.
  # @param color [String] The color of the bike.
  # @param max_speed [Integer] The maximum speed of the bike.
  # @param number_of_tires [Integer] The number of tires of the bike.
  def initialize(color, max_speed, number_of_tires)
    super(color, max_speed, number_of_tires)
    @cadence = 0
    @gear = 1
  end

  # Sets the gear of the bike.
  # @param gear [Integer] The gear to set.
  def gear_set(gear)
    @gear = gear
  end

  # Sets the cadence of the bike.
  # @param cadence [Integer] The cadence to set.
  def cadence_set(cadence)
    @cadence = cadence
  end

  # Accelerates the bike based on cadence and gear.
  # @param applied_power [Integer] The power level for acceleration.
  def accelerate(applied_power)
    @cadence += applied_power
    @current_speed += (@cadence * @gear)
  end

  # Rings the bike's bell.
  # @return [String] The sound of the bell.
  def ring_bell
    "Ding ding!\n"
  end

  # Displays the status of the bike, including cadence and gear.
  def status
    super()
    puts "-> Cadence: #{@cadence}, Gear: #{@gear}"
  end
end

File: 2-I0/2-10-TS/Bike.ts
/*
 * Class representing a bike, extending the Vehicle class.
 * @version 1.0
 * @since   2024-11-05
 */

import Vehicle from "./Vehicle.ts";

export class Bike extends Vehicle {

    /**
     * The cadence of the bike.
     * @type {number}
     */
    protected cadence: number = 0;

    /**
     * The gear of the bike.
     * @type {number}
     */
    protected gear: number = 1;

    /**
     * Creates a new Bike instance.
     * @param {string} color - The color of the bike.
     * @param {number} maxSpeed - The maximum speed of the bike.
     * @param {number} numberOfTires - The number of tires the bike has.
     */
    constructor(color: string, maxSpeed: number, numberOfTires: number) {
        super(color, maxSpeed, numberOfTires);
    }

    /**
     * Sets the gear of the bike.
     * @param {number} gear - The gear to set.
     */
    public setGear(gear: number): void {
        this.gear = gear;
    }

    /**
     * Sets the cadence of the bike.
     * @param {number} cadence - The cadence to set.
     */
    public setCadence(cadence: number): void {
        this.cadence = cadence;
    }

    /**
     * Accelerates the bike based on cadence and gear.
     * @param {number} appliedPower - The power level for acceleration.
     */
    public accelerate(appliedPower: number): void {
        this.cadence = this.cadence + appliedPower;
        this.currentSpeed = this.currentSpeed + (this.cadence * this.gear);
    }

    /**
     * Rings the bike's bell.
     * @returns {string} The sound of the bell.
     */
    public ringBell(): string {
        return "Ding ding!\n";
    }

    /**
     * Displays the status of the bike, including cadence and gear.
     */
    public status(): void {
        super.status();
        console.log(`-> Cadence: ${this.cadence}`);
        console.log(`-> Gear: ${this.gear}`);
    }
}
